test/e2e/dns.go

```diff
@@ -0,0 +1,409 @@
+package e2e
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"regexp"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	"github.com/Azure/go-autorest/autorest/to"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/network"
+	"github.com/Azure/ARO-RP/pkg/util/stringutils"
+	"github.com/Azure/ARO-RP/test/util/project"
+)
+
+var (
+	nameserverRegex              = regexp.MustCompile("nameserver [0-9.]*")
+	verifyResolvConfTimeout      = 30 * time.Second
+	verifyResolvConfPollInterval = 1 * time.Second
+	nodesReadyPollInterval       = 2 * time.Second
+	nicUpdateWaitTime            = 5 * time.Second
+)
+
+const (
+	maxObjNameLen           = 63
+	resolvConfContainerName = "read-resolv-conf"
+)
+
+var _ = Describe("ARO cluster DNS", func() {
+	It("must not be adversely affected by Azure host servicing", func(ctx context.Context) {
+		By("creating a test namespace")
+		testNamespace := fmt.Sprintf("test-e2e-%d", GinkgoParallelProcess())
+		p := project.NewProject(clients.Kubernetes, clients.Project, testNamespace)
+		err := p.Create(ctx)
+		Expect(err).NotTo(HaveOccurred(), "Failed to create test namespace")
+
+		By("verifying the namespace is ready")
+		Eventually(func(ctx context.Context) error {
+			return p.Verify(ctx)
+		}).WithContext(ctx).WithTimeout(DefaultEventuallyTimeout).Should(BeNil())
+
+		DeferCleanup(func(ctx context.Context) {
+			By("deleting the test namespace")
+			err := p.Delete(ctx)
+			Expect(err).NotTo(HaveOccurred(), "Failed to delete test namespace")
+
+			By("verifying the namespace is deleted")
+			Eventually(func(ctx context.Context) error {
+				return p.VerifyProjectIsDeleted(ctx)
+			}).WithContext(ctx).WithTimeout(DefaultEventuallyTimeout).Should(BeNil())
+		})
+
+		By("listing all cluster nodes to retrieve the names of the worker nodes")
+		workerNodes := map[string]string{}
+		nodeList, err := clients.Kubernetes.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		for _, node := range nodeList.Items {
+			name := node.ObjectMeta.Name
+			if strings.Contains(name, "worker") {
```

Comment:
We should try to use the node role / label here, if possible, to avoid possible flakes from the node name.